src/main/java/uk/gov/ida/verifyserviceprovider/services/UnifyingKeySignatureTrustEngine.java

```diff
@@ -0,0 +1,105 @@
+package uk.gov.ida.verifyserviceprovider.services;
+
+import com.google.common.collect.Lists;
+import net.shibboleth.utilities.java.support.resolver.CriteriaSet;
+import net.shibboleth.utilities.java.support.resolver.ResolverException;
+import org.opensaml.security.credential.Credential;
+import org.opensaml.security.credential.CredentialResolver;
+import org.opensaml.xmlsec.keyinfo.KeyInfoCredentialResolver;
+import org.opensaml.xmlsec.signature.support.impl.ExplicitKeySignatureTrustEngine;
+import uk.gov.ida.saml.metadata.EidasMetadataResolverRepository;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.LinkedList;
+import java.util.List;
+
+public class UnifyingKeySignatureTrustEngine extends ExplicitKeySignatureTrustEngine {
+
+    public UnifyingKeySignatureTrustEngine(ExplicitKeySignatureTrustEngine hubTrustEngine, EidasMetadataResolverRepository eidasMetadataResolverRepository) {
```

Comment:
I'm not sure about this class - its sole purpose it to keep a reference the two inner classes. We could just keep those (in their own files) an initialise `ExplicitKeySignatureTrustEngine` directly. It's difficult to follow the logic with classes embedded like this.

Also what's the difference between the two inner classes? They seem to be doing the same thing. Can we refactor that to have less duplication?

Reply:
I see your point about initialising an `ExplicitKeySignatureTrustEngine` in a special method. To my mind it's clearer this way as the class describes what it does, and contains its internal supporting classes - so that you don't need to know about a special way to construct this behaviour if you need it elsewhere, and the internal classes (which aren't ever needed elsewhere) are contained.

The inner classes are instances of a `KeyInfoCredentialResolver` and a `CredentialResolver`. - which I believe are slightly different under the bonnet. Although my implementation of the abstract methods are very similar, I believe we do need both extensions of these classes in order to be able to perform their function against both the Verify and Eidas signature sources.

I could break those sub-classes out into separate class files, but as they're exclusively used by this particular class, I'd recommend they stay there.

Other than by providing a static method that each abstract method override calls, I can't see a helpful way to refactor and reduce the duplication much further - and as it seemed like a small amount of code duplication I felt like it was clearer to have it as is. Happy to change that if you disagree.

Reply:
Done 👍